### PR TITLE
swappy: Init at 1.2.1

### PR DESCRIPTION
**Pull request by matthiasbeyer**
WIP because waiting for

* https://github.com/jtheoof/swappy/issues/10
* https://github.com/jtheoof/swappy/issues/11

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/swappy/default.nix [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  version = "2020-02-26";
+
+  src = fetchFromGitHub {
+    owner = "jtheoof";
+    repo = "swappy";
+    rev = "b5cc433d75d77759cef139e0e232bde79196f886";
+    sha256 = "08r9hmhzzb5ac4g6zwm7w05n99v0hl2h0w4d8i694hg4pyjxb95y";
+  };
+
+  buildInputs = [ meson ninja pkgconfig cmake scdoc gio-sharp libnotify gtk pango cairo wayland];
```
Comment (doronbehar) [inline pkgs/applications/misc/swappy/default.nix]: So upstream [said](https://github.com/jtheoof/swappy/issues/11#issuecomment-591671337): 

> gio-sharp does not exist in Arch.

Are you sure `gio-sharp` is absolutely needed? How come he hasn't said so?

File: pkgs/applications/misc/swappy/default.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  meta = {
+    homepage = https://github.com/jtheoof/swappy;
```
Comment (doronbehar) [inline pkgs/applications/misc/swappy/default.nix]:
```suggestion
    homepage = "https://github.com/jtheoof/swappy";
```

File: pkgs/applications/misc/swappy/default.nix [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+  src = fetchFromGitHub {
+    owner = "jtheoof";
+    repo = "swappy";
+    rev = "b5cc433d75d77759cef139e0e232bde79196f886";
+    sha256 = "08r9hmhzzb5ac4g6zwm7w05n99v0hl2h0w4d8i694hg4pyjxb95y";
+  };
+
+  buildInputs = [ meson ninja pkgconfig cmake scdoc gio-sharp libnotify gtk pango cairo wayland];
+
```
Comment (doronbehar) [inline pkgs/applications/misc/swappy/default.nix]: As for https://github.com/jtheoof/swappy/issues/10 , try to add (untested):

```
  nativeBuildInputs = [ glibc.dev ];
```

Source: https://linux.die.net/man/3/clock_gettime + `nix-locate unistd.h`.

**Review by doronbehar — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/swappy/default.nix (hunk 1 above)
Comment (matthiasbeyer) [reply pkgs/applications/misc/swappy/default.nix]: I get errors without it, e.g.:

```
../src/pixbuf.c:3:10: fatal error: gio/gunixoutputstream.h: No such file or directory
    3 | #include <gio/gunixoutputstream.h>
```

File: pkgs/applications/misc/swappy/default.nix (hunk 3 above)
Comment (matthiasbeyer) [reply pkgs/applications/misc/swappy/default.nix]: Did not work:

```
../src/buffer.c: In function 'randname':
../src/buffer.c:12:3: error: implicit declaration of function 'clock_gettime' [-Werror=implicit-function-declaration]
   12 |   clock_gettime(CLOCK_REALTIME, &ts);
      |   ^~~~~~~~~~~~~
../src/buffer.c:12:17: error: 'CLOCK_REALTIME' undeclared (first use in this function)
   12 |   clock_gettime(CLOCK_REALTIME, &ts);
      |                 ^~~~~~~~~~~~~~
../src/buffer.c:12:17: note: each undeclared identifier is reported only once for each function it appears in
../src/buffer.c: In function 'create_shm_file':
../src/buffer.c:45:7: error: implicit declaration of function 'ftruncate'; did you mean 'strncat'? [-Werror=implicit-function-declaration]
   45 |   if (ftruncate(fd, size) < 0) {
      |       ^~~~~~~~~
      |       strncat
cc1: all warnings being treated as errors
```

File: pkgs/applications/misc/swappy/default.nix (hunk 1 above)
Comment (jtheoof) [reply pkgs/applications/misc/swappy/default.nix:29]: It's part of `glib2` on Arch. Somehow it compiles fine on Arch without specifying the lib explicitly. I just pushed a fix in the `meson` build file.

File: pkgs/applications/misc/swappy/default.nix (hunk 3 above)
Comment (jtheoof) [reply pkgs/applications/misc/swappy/default.nix:30]: This should be fixed now too.

**Comment by jtheoof** [comment]
The latest build from `master` should compile better. Let me know if you're still having issues. Any help is appreciated too, I don't have NixOS installed.

**Comment by bbigras** [comment]
1.0.0 is out. https://github.com/jtheoof/swappy/releases

@matthiasbeyer don't hesitate to say if you don't have the time for this PR anymore. One of us could pick it up.

**Comment by matthiasbeyer** [comment]
No, I'm fine. I pushed an update to 1.0.0, but it still fails to compile because of `gio` and I'm failing to find a fix.

**Comment by jtheoof** [comment]
Can't really help you, I don't have nixOS installed. On Arch, `gio.h` is part of `glib2`.

File: pkgs/applications/misc/swappy/default.nix [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    owner = "jtheoof";
+    repo = "swappy";
+    rev = "v${version}";
+    sha256 = "1dk5q1f71n5zs2xximpwi2jfdznqwvjrprxi7clqc077nqv48s1j";
+  };
+
+  nativeBuildInputs = [ glibc.dev ];
+
+  buildInputs = [ meson ninja pkgconfig cmake scdoc gio-sharp libnotify gtk pango cairo wayland];
+
```
Comment (bbigras) [inline pkgs/applications/misc/swappy/default.nix:30]: It builds with this change:

```suggestion

  preConfigure = ''
    sed -i "s#gio/gunixoutputstream.h#gio-unix-2.0/gio/gunixoutputstream.h#" src/pixbuf.c
  '';

```

Comment (bbigras) [reply pkgs/applications/misc/swappy/default.nix:30]: Or maybe there's a way to add `${glib.dev}/include/gio-unix-2.0` to the include path.

Comment (bbigras) [reply pkgs/applications/misc/swappy/default.nix:30]: Just use:
```nix
  mesonFlags = [
    # TODO: https://github.com/NixOS/nixpkgs/issues/36468
    "-Dc_args=-I${glib.dev}/include/gio-unix-2.0"
  ];
```

and you can remove the `nativeBuildInputs`

Comment (bbigras) [reply pkgs/applications/misc/swappy/default.nix:30]: and you can remove `gio-sharp`.

**Comment by jtheoof** [comment]
There was a recent patch in swappy: https://github.com/jtheoof/swappy/pull/42 that changes slightly how `gio` is found during `meson` configuration. It might help you a little.

**Comment by bbigras** [comment]
https://github.com/jtheoof/swappy/commit/606ce38914f3735ca0aad0fc982030130ff95d65 has the same problem.

**Comment by bbigras** [comment]
@matthiasbeyer friendly ping

**Comment by matthiasbeyer** [comment]
Not sure what I should do here.

**Comment by bbigras** [comment]
try adding:

```nix
  mesonFlags = [
    # TODO: https://github.com/NixOS/nixpkgs/issues/36468
    "-Dc_args=-I${glib.dev}/include/gio-unix-2.0"
  ];
```

You can check the URL for more info.

**Comment by matthiasbeyer** [comment]
Fixed thanks to @bbigras and #36468 and also updated to release 1.2.0.

File: pkgs/applications/misc/swappy/default.nix [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  src = fetchFromGitHub {
+    owner = "jtheoof";
+    repo = "swappy";
+    rev = "v${version}";
+    sha256 = "13ky6nx6lprsiza62syqrq63i78fcg74lzd9pwa4kabjnqc350j5";
+  };
+
+  nativeBuildInputs = [ glibc.dev ];
+
+  buildInputs = [ meson ninja pkgconfig cmake scdoc gio-sharp libnotify gtk pango cairo wayland];
```
Comment (bbigras) [inline pkgs/applications/misc/swappy/default.nix:30]: > build time only dependencies are declared in `nativeBuildInputs`
https://nixos.org/nixpkgs/manual/#reviewing-contributions-new-packages

So I guess at least: meson, ninja, pkgconfig, cmake. Not sure about the others.

and it looks like "glibc.dev" can be removed;

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -2235,6 +2235,8 @@ in
 
   bash-supergenpass = callPackage ../tools/security/bash-supergenpass { };
 
+  swappy = callPackage ../applications/misc/swappy { gtk = gtk3; };
```
Comment (bbigras) [inline pkgs/top-level/all-packages.nix:2238]: flameshot, a similar tool is in pkgs/tools/misc/flameshot. Not sure if it's important.

File: pkgs/applications/misc/swappy/default.nix [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  mesonFlags = [
+    # TODO: https://github.com/NixOS/nixpkgs/issues/36468
+    "-Dc_args=-I${glib.dev}/include/gio-unix-2.0"
+  ];
+
+  meta = {
+    homepage = "https://github.com/jtheoof/swappy";
+    description = "A Wayland native snapshot editing tool, inspired by Snappy on macOS ";
```
Comment (bbigras) [inline pkgs/applications/misc/swappy/default.nix:39]: maybe remove the whitespace at the end

**Comment by bbigras** [comment]
it seems `gio-sharp` can be removed too.

I was able to build using:

```nix
  nativeBuildInputs = [ glib meson ninja pkgconfig cmake scdoc ];
  buildInputs = [ cairo pango gtk libnotify wayland glib ];
  strictDeps = true;
```

Not sure why I need `glib` in both places.

**Comment by bbigras** [comment]
@matthiasbeyer friendly ping

**Comment by matthiasbeyer** [comment]
With your suggested change:

```
[16/18] Compiling C object 'swappy@exe/src_pixbuf.c.o'.o'.ppy.c.o'.
FAILED: swappy@exe/src_pixbuf.c.o
/nix/store/1kn7fi3hhi33jms3113riyzwyn2yqpqd-gcc-wrapper-9.2.0/bin/cc -Iswappy@exe -I. -I.. -I../include -Ires -I/nix/store/a92yhf0icjl9lh553s4qjchdsm8lyzsy-cairo-1.16.0-dev/include/cairo -I/nix/store/1zlkr4x0bvxw3mn65chw9v3b48n8jh52-freetype-2.10.1-dev/include/freetype2 -I/nix/store/1zlkr4x0bvxw3mn65chw9v3b48n8jh52-freetype-2.10.1-dev/include -I/nix/store/5ad0cyv6lxpkk2m1prcrasdp2pysz73q-glib-2.62.4-dev/include -I/nix/store/5ad0cyv6lxpkk2m1prcrasdp2pysz73q-glib-2.62.4-dev/include/glib-2.0 -I/nix/store/vwvjrh97byihk40gwl26f0lnb3w405cv-glib-2.62.4/lib/glib-2.0/include -I/nix/store/sm843466f7cvhf1b804jjlay1nv7fyq0-pango-1.44.7-dev/include/pango-1.0 -I/nix/store/70yjv9sr9mhf75bliwmmxgyr37mx9lac-harfbuzz-2.6.4-dev/include/harfbuzz -I/nix/store/i5yw4gdhknqahzhkaqjfbgc9ydb2885w-gtk+3-3.24.13-dev/include/gtk-3.0 -I/nix/store/vmm5hrcbqrb7b605dnir17qbiz85x17n-atk-2.34.1-dev/include/atk-1.0 -I/nix/store/4w16ai3cx8sp7av9kc432w34r4fpdmfp-gdk-pixbuf-2.40.0-dev/include/gdk-pixbuf-2.0 -I/nix/store/i1mssjnbih9czf12fcda1883pvkirgss-libnotify-0.7.8-dev/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Werror -std=c11 -Wno-unused-parameter -DHAVE_LIBNOTIFY -MD -MQ 'swappy@exe/src_pixbuf.c.o' -MF 'swappy@exe/src_pixbuf.c.o.d' -o 'swappy@exe/src_pixbuf.c.o' -c ../src/pixbuf.c
../src/pixbuf.c:4:10: fatal error: gio/gunixoutputstream.h: No such file or directory
    4 | #include <gio/gunixoutputstream.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[17/18] Compiling C object 'swappy@exe/src_render.c.o'.
ninja: build stopped: subcommand failed.
builder for '/nix/store/v3frm48zb0pp5j2r74g1sh0h5s1maacg-swappy-1.0.0.drv' failed with exit code 1
error: build of '/nix/store/v3frm48zb0pp5j2r74g1sh0h5s1maacg-swappy-1.0.0.drv' faile
```

**Comment by bbigras** [comment]
@GrahamcOfBorg build swappy

**Comment by bbigras** [comment]
##### Reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on x86_64
- [x] executables tested on x86_64
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] phases are respected
- [x] patches that are remotely available are fetched with `fetchpatch`

##### Possible improvements

##### Comments

Other reviewers might want to double check `nativeBuildInputs` as i'm not an expert on that.

**Review by bbigras — APPROVED** [review]

**Comment by blaggacao** [comment]
Just noticed v1.2.1 was cut 14 days ago.
https://github.com/jtheoof/swappy/tree/v1.2.1

Not meaning to block anything.

File: pkgs/applications/misc/swappy/default.nix [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    license = stdenv.lib.licenses.mit;
+    maintainers = [ stdenv.lib.maintainers.matthiasbeyer ];
+    platforms = with stdenv.lib.platforms; linux;
```
Comment (jonringer) [inline pkgs/applications/misc/swappy/default.nix:43]: personal preference:
```suggestion
  meta = stdenv.lib; {
    homepage = "https://github.com/jtheoof/swappy";
    description = "A Wayland native snapshot editing tool, inspired by Snappy on macOS ";
    license = licenses.mit;
    maintainers = [ maintainers.matthiasbeyer ];
    platforms = with platforms; linux;
```

**Comment by bbigras** [comment]
@matthiasbeyer friendly ping

**Comment by jonringer** [comment]
since this is the first time it's been packaged, I think it should be a single commit. I don't see much value in having 5 commits.

**Comment by bbigras** [comment]
@matthiasbeyer friendly ping. Can we get this in the release please?

**Comment by matthiasbeyer** [comment]
Sure, what's missing from your point of view?

**Comment by bbigras** [comment]
I already approved but maybe Jon's meta suggestion and squash the commits and it will be good to go.

**Comment by bbigras** [comment]
@jonringer can you merge this please?

File: pkgs/applications/misc/swappy/default.nix (hunk 7 above)
Comment (AndersonTorres) [reply pkgs/applications/misc/swappy/default.nix:40]: @matthiasbeyer cleanup whitespace

**Comment by berbiche** [comment]
Friendly ping @jonringer since two weeks have passed since the last one.

I'm looking forward to using this new tool in my workflow.

It would also be great if this package was backported to the 20.09 release (unless the window is closed?).

**Comment by bbigras** [comment]
@marsam help merge this please. Jon must be pretty busy these days.

**Review by marsam — APPROVED** [review]

**Comment by bbigras** [comment]
Thank you very much :smile:

**Pull request merged**